### ISSUE TITLE
New AdjustedCount converter function

### ISSUE THREAD

**Issue by lahsivjar** [open-telemetry/opentelemetry-collector-contrib#36572]
### Component(s)

pkg/ottl

### Is your feature request related to a problem? Please describe.

[Adjusted count](https://opentelemetry.io/docs/specs/otel/trace/tracestate-probability-sampling-experimental/#adjusted-count) provides a measure of how many spans an individually sampled span represents. It would be nice to have a function in OTTL that could calculate the adjusted count for a span.

### Describe the solution you'd like

Add a new `AdjustedCount` converter function. The function will be applicable only for spans (similar to [IsRootSpan](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md#isrootspan) function). The adjusted count can be calculated by using the [sampling pkg](https://pkg.go.dev/github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling), something like:

```go
import (
	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling"
)

func CalculateAdjustedCount(tracestate string) float64 {
	w3cTraceState, err := sampling.NewW3CTraceState(tracestate)
	if err != nil {
		return 1
	}
	otTraceState := w3cTraceState.OTelValue()
	if otTraceState == nil {
		return 1
	}
	if len(otTraceState.TValue()) == 0 {
		// For non-probabilistic sampler OR always sampling threshold, default to 1
		return 1
	}
	return otTraceState.AdjustedCount()
}
```

**Comment by lahsivjar** [comment]
Created a PR to show how this looks like: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36573

**Comment by TylerHelmuth** [comment]
@lahsivjar OTTL has access to the trace_state, is there a way to calculate this value using the trace_state and OTTL's arithmetic capabilities?

If not adding support for this otel feature is appropriate.  I wonder tho if a converter or a new path segment is better. We could add to the `ottlspan` context the idea of `trace_state.adjusted_count` to get the value.

@evan-bradley what do you think?

**Comment by evan-bradley** [comment]
I agree, `trace_state.adjusted_count` makes more sense since there are no parameters and we're essentially getting a modified version of a specific path.

At a quick glance of the `pkg/sampling` package the PR uses, the functionality for the adjusted count also appears to be something that is better handled by a library instead of in OTTL, so I think it's okay to add an explicit path for it.

**Comment by lahsivjar** [comment]
> I believe it's worth having, and should use pkg/sampling, as the math is slightly tricky. Even if OTTL could do it, this is important to get right.

Agreed on using the pkg/sampling, I initially tried using OTTL expression to do this but in addition to it being complicated to reason/maintain the OTTL expression can get outdated if the spec around sampling evolves.

> I agree, trace_state.adjusted_count makes more sense since there are no parameters and we're essentially getting a modified version of a specific path.

Introducing this as a path also sounds good to me, and, IIANM, we could still use the pkg/sampling to do this.

> OTEP 235 has not landed yet at the W3C, so the spec isn't completely finalized.
The API in pkg/sampling is therefore not yet final (but it's probably pretty close).

How do we deal with this? Are there any concerns about evolving/introducing breaking changes in the OTTL implementation if the spec changes?

> If we're going to introduce AdjustedCount we should also consider introducing at least SamplingProbability and possibly SamplingThreshold at the same time, since people seem to have strong preferences as to which they need.

ACK, I have planned to work on at least the SamplingProbability as a follow-up but I can also include the SamplingThreshold. Going back to the implementation, should we keep these as converters, or are paths more suitable? I don't have a strong inclination either way but I don't have in-depth knowledge about OTTL grammar -- the points presented in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36572#issuecomment-2521021999 sound reasonable to me though.

**Comment by evan-bradley** [comment]
>Going back to the implementation, should we keep these as converters, or are paths more suitable?

I would still like to stick to implementing these as paths. Converters are usually for converting arbitrary input data to some output format. In this case, we're looking to access data on a specific location on the span.

**Comment by lahsivjar** [comment]
> I would still like to stick to implementing these as paths

I have a draft PR for adding `trace_state.adjusted_count` but I am having second thoughts on the path approach. AFAIU, the paths seem to have both getters and setters and it makes sense since they usually refer directly to something that can be accessed. However, in the case of adjusted count, it is a derived value from a trace state and, given an adjusted count, we cannot update the trace state (at least based on my understanding but CMIIW).

**Comment by TylerHelmuth** [comment]
We support some other get-only paths (metric type), it is ok if we don't implement a set.

**Comment by jmacd** [comment]
In addition to OTEP 235, the `probabilisticsampler` processor adds a (draft) semantic conventions for log records that I think we should follow. We are not waiting on W3C, and OTEP 235 was merged at this point: after https://github.com/open-telemetry/opentelemetry-specification/pull/4166 merges, the ship sails. There are two low-level functions with different logic depending on the context. 

- randomness(): defined in https://github.com/open-telemetry/opentelemetry-specification/pull/4162
- threshold(): defined in https://github.com/open-telemetry/opentelemetry-specification/pull/4166

I suspect that `adjusted_count()` can be derived from OTTL arithmetic expressions over `threshold()`, but it's probably worth making adjusted count easy to compute. The sampling decision equals `randomness() >= threshold()` -- I'd make all three of these available as builtins.

Threshold:
- For span contexts, the tracestate ot=th:X value is used.
- For log contexts, the record sampling.threshold attribute is used.

Randomness:
- For span contexts, the tracestate ot=rv:X or TraceID[9:16] is used.
- For log contexts, the sampling.randomness attribute is used.
